### PR TITLE
SDL320

### PR DESCRIPTION
**Pull request by mbrookhart**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/contrib/ngraph.cc [hunk 1]
```diff
@@ -180,6 +184,10 @@ bool NgraphSubgraphInferShape(const nnvm::NodeAttrs &attrs,
 bool NgraphSubgraphInferType(const nnvm::NodeAttrs &attrs,
                              std::vector<int> *iattr, std::vector<int> *oattr) {
   auto graph = get_ngraph(attrs);
+  CHECK(iattr != nullptr);
+  CHECK(oattr != nullptr);
+  CHECK(iattr->size() == graph->inputs_.size());
+  CHECK(oattr->size() == graph->get_results().size());
```
Comment (cconvey) [inline src/operator/contrib/ngraph.cc]: Would it make sense to move the checks regarding `oattr` further down, closer to where `oattr` is actually used?  I think that would better adhere to the coding style you've used elsewhere for where the checks are added.

File: src/operator/contrib/ngraph.cc [hunk 2]
```diff
@@ -198,6 +206,9 @@ bool NgraphSubgraphInferStorageType(const nnvm::NodeAttrs &attrs,
                                     mxnet::DispatchMode *dispatch_mode,
                                     std::vector<int> *in_attrs,
                                     std::vector<int> *out_attrs) {
+  CHECK(dispatch_mode != nullptr);
+  CHECK(in_attrs != nullptr);
+  CHECK(out_attrs != nullptr);
```
Comment (cconvey) [inline src/operator/contrib/ngraph.cc]: Slightly off-topic, but can we add curly braces to the body of the `if` statement below?  Their absence led me to initially mis-read the code.

File: src/operator/contrib/ngraph.cc [hunk 3]
```diff
@@ -161,6 +162,12 @@ bool NgraphSubgraphInferShape(const nnvm::NodeAttrs &attrs,
   auto compiler =
       nnvm::get<std::shared_ptr<ngraph_bridge::Compiler>>(attrs.parsed);
   auto graph = get_ngraph(attrs);
+
+  ngraph_bridge::check(in_attrs != nullptr);
+  ngraph_bridge::check(out_attrs != nullptr);
+  ngraph_bridge::check(in_attrs->size() == graph->inputs_.size());
+  ngraph_bridge::check(out_attrs->size() == graph->get_results().size());
```
Comment (cconvey) [inline src/operator/contrib/ngraph.cc]: Why do we use `ngraph_bridge::check` here but `CHECK` in some other places?

File: src/ndarray/ndarray.cc [hunk 4]
```diff
@@ -746,20 +746,20 @@ std::shared_ptr<ngraph::runtime::Tensor> &NDArray::create_tensor(
       ptr_->tensor_view_->get_shape() !=
           ngraph_bridge::TShape_to_NShape(shape_)) {
     auto backend = ngraph_bridge::GetBackendFromContext(ctx());
+    void* dptr = storage_handle().dptr;
     CHECK(backend != nullptr);
```
Comment (louisfeng) [inline src/ndarray/ndarray.cc]: I think if we already have a macro around these code, we could use `ngraph_bridge::check()` here like the rest of the code instead of `CHECK()`. Although, I don't know why we created a our own check in the first place rather than using the mxnet version.

Comment (mbrookhart) [reply src/ndarray/ndarray.cc]: !!! Will fix. To answer your question: 

Klockwork doesn’t like MXNet’s `CHECK` macro, complains about various memory violations. When I added the negative test cases and actually hit one of those checks, instead of throwing an error that GTEST could catch and validate, `CHECK` returned `SIGABRT` or `SIGSEGV`. Rather than trying to fix the rabbit hole of macros, I just wrote our own utility and replaced all of the instances we’re using in the bridge.
